File: web/src/api/drama.ts
import { http } from "./client";

// 追剧转存任务 API 封装（来自Trae）。
// 对应后端 internal/api/drama.go 的 /api/admin/drama/* 路由。

// DramaTask 追剧转存任务（来自Trae）
export interface DramaTask {
  id: number;
  task_name: string;
  account_id: number;
  share_url: string;
  save_path: string;
  pattern: string;
  replace: string;
  ignore_extension: boolean;
  run_week: string; // 逗号分隔的 1-7
  end_date: string; // YYYY-MM-DD
  update_subdir: string;
  update_subdir_resave_mode: string;
  start_fid: string;
  sort_index: number;
  status: string; // running/paused
  last_run_at?: string;
  last_run_status?: string;
  last_run_message?: string;
  last_tree_summary?: string;
  created_at?: string;
  updated_at?: string;
}

// DramaTaskInput 创建/更新任务入参（来自Trae）
export interface DramaTaskInput {
  task_name: string;
  account_id: number;
  share_url: string;
  save_path: string;
  pattern: string;
  replace: string;
  ignore_extension: boolean;
  run_week: string;
  end_date: string;
  update_subdir: string;
  update_subdir_resave_mode: string;
  start_fid: string;
  sort_index: number;
  status: string;
}

// DramaTaskRun 执行历史（来自Trae）
export interface DramaTaskRun {
  id: number;
  task_id: number;
  status: string; // running/success/failed/skipped
  message?: string;
  tree_summary?: string;
  transfer_count: number;
  started_at?: string;
  finished_at?: string;
}

// RunDramaTaskResult 手动触发执行结果（来自Trae）
export interface RunDramaTaskResult {
  submitted: boolean;
  task_id: number;
  allow_once: boolean;
  message: string;
}

export function fetchDramaTasks() {
  return http.get<DramaTask[]>("/admin/drama/tasks");
}

export function createDramaTask(body: DramaTaskInput) {
  return http.post<DramaTask>("/admin/drama/tasks", body);
}

export function getDramaTask(id: number) {
  return http.get<DramaTask>(`/admin/drama/tasks/${id}`);
}

export function updateDramaTask(id: number, body: DramaTaskInput) {
  return http.put<DramaTask>(`/admin/drama/tasks/${id}`, body);
}

export function deleteDramaTask(id: number) {
  return http.del<{ id: number }>(`/admin/drama/tasks/${id}`);
}

// runDramaTask 手动触发执行一次（来自Trae）。
// once=true 跳过星期/截止校验；默认 true。
export function runDramaTask(id: number, once = true) {
  return http.post<RunDramaTaskResult>(`/admin/drama/tasks/${id}/run`, {}, { once });
}

export function fetchDramaTaskRuns(id: number, limit = 20) {
  return http.get<DramaTaskRun[]>(`/admin/drama/tasks/${id}/runs`, { limit });
}

// === 命名正则规则维护（来自Trae，对应 /admin/drama/rules） ===

// MagicRegexRule 命名正则规则（来自Trae）
export interface MagicRegexRule {
  key: string; // 以 $ 开头的规则键
  label: string; // 展示名称
  enabled: boolean; // 是否启用
  built_in: boolean; // 是否为内置规则
  overridden: boolean; // 内置规则是否被覆盖
  pattern: string; // 匹配正则
  replace: string; // 替换模板
  default_pattern?: string; // 内置默认 pattern（仅内置规则）
  default_replace?: string; // 内置默认 replace（仅内置规则）
}

// MagicRegexRuleInput 规则写入入参（来自Trae）
export interface MagicRegexRuleInput {
  label?: string | null;
  enabled?: boolean;
  pattern?: string | null;
  replace?: string | null;
}

// fetchMagicRegexRules 获取全部规则（内置 + 自定义）（来自Trae）
export function fetchMagicRegexRules() {
  return http.get<{ rules: MagicRegexRule[] }>("/admin/drama/rules");
}

// saveMagicRegexRule 创建/覆盖规则（来自Trae），返回更新后的规则列表
export function saveMagicRegexRule(key: string, body: MagicRegexRuleInput) {
  return http.put<{ rules: MagicRegexRule[] }>(`/admin/drama/rules/${encodeURIComponent(key)}`, body);
}

// deleteMagicRegexRule 删除规则（来自Trae），返回更新后的规则列表
export function deleteMagicRegexRule(key: string) {
  return http.del<{ rules: MagicRegexRule[] }>(`/admin/drama/rules/${encodeURIComponent(key)}`);
}

// === 分享链接预览（来自Trae，对应 /admin/drama/share/preview） ===

// SharePreviewItem 分享预览条目（来自Trae）
export interface SharePreviewItem {
  fid: string;
  fid_token?: string;
  name: string; // 原始文件名
  name_re?: string; // 重命名后的目标名；为空表示不参与转存
  is_dir: boolean;
  updated_at: number;
  size: number;
  children_count: number;
  name_saved?: string; // 状态标注：已存在文件名 / 起始及之前 / 重命名冲突
}

// SharePreviewResult 分享预览结果（来自Trae）
export interface SharePreviewResult {
  drive_type: string;
  pwd_id: string;
  pdir_fid: string;
  items: SharePreviewItem[];
}

// SharePreviewInput 预览入参（来自Trae）
export interface SharePreviewInput {
  account_id: number;
  share_url: string;
  pdir_fid?: string; // 浏览层级用；空则从分享链接解析
  max_items?: number;
  task_name?: string;
  pattern?: string;
  replace?: string;
  sort_index?: number;
  save_path?: string;
  ignore_extension?: boolean;
  update_subdir?: string;
  start_fid?: string;
}

// previewShare 预览分享目录内容与重命名效果（来自Trae）
export function previewShare(body: SharePreviewInput) {
  return http.post<SharePreviewResult>("/admin/drama/share/preview", body);
}

File: internal/api/drama.go
package api

// 追剧转存任务的 HTTP 处理器（来自Trae）。
// 复用 ensureServiceReady / decodeJSON / pathID / writeOK / writeErr 等通用助手，
// 与 automation.go 的风格保持一致。

import (
	"net/http"
	"strconv"

	"litepan/internal/domain"
)

// dramaTaskDTO 追剧任务的对外响应结构（来自Trae）
type dramaTaskDTO struct {
	ID                     int64  `json:"id"`
	TaskName               string `json:"task_name"`
	AccountID              int64  `json:"account_id"`
	ShareURL               string `json:"share_url"`
	SavePath               string `json:"save_path"`
	Pattern                string `json:"pattern"`
	Replace                string `json:"replace"`
	IgnoreExtension        bool   `json:"ignore_extension"`
	RunWeek                string `json:"run_week"`
	EndDate                string `json:"end_date"`
	UpdateSubdir           string `json:"update_subdir"`
	UpdateSubdirResaveMode string `json:"update_subdir_resave_mode"`
	StartFID               string `json:"start_fid"`
	SortIndex              int    `json:"sort_index"`
	Status                 string `json:"status"`
	LastRunAt              string `json:"last_run_at,omitempty"`
	LastRunStatus          string `json:"last_run_status,omitempty"`
	LastRunMessage         string `json:"last_run_message,omitempty"`
	LastTreeSummary        string `json:"last_tree_summary,omitempty"`
	CreatedAt              string `json:"created_at,omitempty"`
	UpdatedAt              string `json:"updated_at,omitempty"`
}

// dramaTaskInput 创建/更新任务的入参（来自Trae）
type dramaTaskInput struct {
	TaskName               string `json:"task_name"`
	AccountID              int64  `json:"account_id"`
	ShareURL               string `json:"share_url"`
	SavePath               string `json:"save_path"`
	Pattern                string `json:"pattern"`
	Replace                string `json:"replace"`
	IgnoreExtension        bool   `json:"ignore_extension"`
	RunWeek                string `json:"run_week"`
	EndDate                string `json:"end_date"`
	UpdateSubdir           string `json:"update_subdir"`
	UpdateSubdirResaveMode string `json:"update_subdir_resave_mode"`
	StartFID               string `json:"start_fid"`
	SortIndex              int    `json:"sort_index"`
	Status                 string `json:"status"`
}

// dramaTaskRunDTO 执行历史响应结构（来自Trae）
type dramaTaskRunDTO struct {
	ID            int64  `json:"id"`
	TaskID        int64  `json:"task_id"`
	Status        string `json:"status"`
	Message       string `json:"message,omitempty"`
	TreeSummary   string `json:"tree_summary,omitempty"`
	TransferCount int    `json:"transfer_count"`
	StartedAt     string `json:"started_at,omitempty"`
	FinishedAt    string `json:"finished_at,omitempty"`
}

func toDramaTaskDTO(t *domain.DramaTask) dramaTaskDTO {
	dto := dramaTaskDTO{
		ID:                     t.ID,
		TaskName:               t.TaskName,
		AccountID:              t.AccountID,
		ShareURL:               t.ShareURL,
		SavePath:               t.SavePath,
		Pattern:                t.Pattern,
		Replace:                t.Replace,
		IgnoreExtension:        t.IgnoreExtension,
		RunWeek:                t.RunWeek,
		EndDate:                t.EndDate,
		UpdateSubdir:           t.UpdateSubdir,
		UpdateSubdirResaveMode: t.UpdateSubdirResaveMode,
		StartFID:               t.StartFID,
		SortIndex:              t.SortIndex,
		Status:                 t.Status,
		LastRunAt:              t.LastRunAt,
		LastRunStatus:          t.LastRunStatus,
		LastRunMessage:         t.LastRunMessage,
		LastTreeSummary:        t.LastTreeSummary,
	}
	if !t.CreatedAt.IsZero() {
		dto.CreatedAt = FormatAPITime(t.CreatedAt)
	}
	if !t.UpdatedAt.IsZero() {
		dto.UpdatedAt = FormatAPITime(t.UpdatedAt)
	}
	return dto
}

func toDramaTaskRunDTO(r *domain.DramaTaskRun) dramaTaskRunDTO {
	dto := dramaTaskRunDTO{
		ID:            r.ID,
		TaskID:        r.TaskID,
		Status:        r.Status,
		Message:       r.Message,
		TreeSummary:   r.TreeSummary,
		TransferCount: r.TransferCount,
	}
	if !r.StartedAt.IsZero() {
		dto.StartedAt = FormatAPITime(r.StartedAt)
	}
	if !r.FinishedAt.IsZero() {
		dto.FinishedAt = FormatAPITime(r.FinishedAt)
	}
	return dto
}

// applyInput 将入参写入领域对象，用于创建/更新（来自Trae）
func applyInput(t *domain.DramaTask, in dramaTaskInput) {
	t.TaskName = in.TaskName
	t.AccountID = in.AccountID
	t.ShareURL = in.ShareURL
	t.SavePath = in.SavePath
	t.Pattern = in.Pattern
	t.Replace = in.Replace
	t.IgnoreExtension = in.IgnoreExtension
	t.RunWeek = in.RunWeek
	t.EndDate = in.EndDate
	t.UpdateSubdir = in.UpdateSubdir
	t.UpdateSubdirResaveMode = in.UpdateSubdirResaveMode
	t.StartFID = in.StartFID
	t.SortIndex = in.SortIndex
	if in.Status != "" {
		t.Status = in.Status
	} else if t.Status == "" {
		t.Status = "running" // 默认启用（来自Trae）
	}
}

func (h *Handler) listDramaTasks(w http.ResponseWriter, r *http.Request) {
	if !ensureServiceReady(w, h.drama != nil) {
		return
	}
	tasks, err := h.drama.ListTasks(r.Context())
	if err != nil {
		writeErr(w, err)
		return
	}
	out := make([]dramaTaskDTO, 0, len(tasks))
	for _, t := range tasks {
		if t == nil {
			continue
		}
		out = append(out, toDramaTaskDTO(t))
	}
	writeOK(w, out)
}

func (h *Handler) createDramaTask(w http.ResponseWriter, r *http.Request) {
	if !ensureServiceReady(w, h.drama != nil) {
		return
	}
	var in dramaTaskInput
	if err := decodeJSON(r, &in); err != nil {
		writeErr(w, err)
		return
	}
	task := &domain.DramaTask{}
	applyInput(task, in)
	id, err := h.drama.CreateTask(r.Context(), task)
	if err != nil {
		writeErr(w, err)
		return
	}
	task.ID = id
	writeOK(w, toDramaTaskDTO(task))
}

func (h *Handler) getDramaTask(w http.ResponseWriter, r *http.Request) {
	if !ensureServiceReady(w, h.drama != nil) {
		return
	}
	id, err := pathID(r)
	if err != nil {
		writeErr(w, err)
		return
	}
	task, err := h.drama.GetTask(r.Context(), id)
	if err != nil {
		writeErr(w, err)
		return
	}
	if task == nil {
		writeErr(w, domain.Errorf(domain.CodeNotFound, "追剧任务不存在"))
		return
	}
	writeOK(w, toDramaTaskDTO(task))
}

func (h *Handler) updateDramaTask(w http.ResponseWriter, r *http.Request) {
	if !ensureServiceReady(w, h.drama != nil) {
		return
	}
	id, err := pathID(r)
	if err != nil {
		writeErr(w, err)
		return
	}
	var in dramaTaskInput
	if err := decodeJSON(r, &in); err != nil {
		writeErr(w, err)
		return
	}
	task, err := h.drama.GetTask(r.Context(), id)
	if err != nil {
		writeErr(w, err)
		return
	}
	if task == nil {
		writeErr(w, domain.Errorf(domain.CodeNotFound, "追剧任务不存在"))
		return
	}
	applyInput(task, in)
	task.ID = id
	if err := h.drama.UpdateTask(r.Context(), task); err != nil {
		writeErr(w, err)
		return
	}
	writeOK(w, toDramaTaskDTO(task))
}

func (h *Handler) deleteDramaTask(w http.ResponseWriter, r *http.Request) {
	if !ensureServiceReady(w, h.drama != nil) {
		return
	}
	id, err := pathID(r)
	if err != nil {
		writeErr(w, err)
		return
	}
	if err := h.drama.DeleteTask(r.Context(), id); err != nil {
		writeErr(w, err)
		return
	}
	writeOK(w, map[string]any{"id": id})
}

// runDramaTask 手动触发执行，默认 allowOnce=true 跳过星期/截止校验（来自Trae）。
// 传入 ?once=false 时按调度规则校验。异步执行，立即返回提交结果。
func (h *Handler) runDramaTask(w http.ResponseWriter, r *http.Request) {
	if !ensureServiceReady(w, h.drama != nil) {
		return
	}
	id, err := pathID(r)
	if err != nil {
		writeErr(w, err)
		return
	}
	allowOnce := true
	if raw := r.URL.Query().Get("once"); raw != "" {
		if v, convErr := strconv.ParseBool(raw); convErr == nil {
			allowOnce = v
		}
	}
	data := h.drama.RunTaskAsync(id, allowOnce)
	writeOK(w, data)
}

func (h *Handler) listDramaTaskRuns(w http.ResponseWriter, r *http.Request) {
	if !ensureServiceReady(w, h.drama != nil) {
		return
	}
	id, err := pathID(r)
	if err != nil {
		writeErr(w, err)
		return
	}
	limit := 20
	if raw := r.URL.Query().Get("limit"); raw != "" {
		got, convErr := strconv.Atoi(raw)
		if convErr != nil {
			writeErr(w, domain.Errorf(domain.CodeValidation, "非法 limit：%s", raw))
			return
		}
		limit = got
	}
	runs, err := h.drama.ListTaskRuns(r.Context(), id, limit)
	if err != nil {
		writeErr(w, err)
		return
	}
	out := make([]dramaTaskRunDTO, 0, len(runs))
	for _, run := range runs {
		if run == nil {
			continue
		}
		out = append(out, toDramaTaskRunDTO(run))
	}
	writeOK(w, out)
}
